src/fifelse.c

```diff
@@ -0,0 +1,338 @@
+#include <R.h>
+#include <Rdefines.h>
+#include <R_ext/Error.h>
+#include<stdint.h>
+
+SEXP fifelseR(SEXP l, SEXP a, SEXP b)
+{
+  // l is the test
+  // a is what to do in case l is TRUE
+  // b is what to do in case l is FALSE
+  // maybe add something in case l is neither TRUE or FALSE ? for now it will default to NA
+  
+  // Check if test is logical
+  if(!isLogical(l)) error("Argument 'test' must be logical.");
+  
+  SEXPTYPE ta = TYPEOF(a); 
+  
+  // Check if same type
+  if(ta != TYPEOF(b))
+    error("Item 'yes' is of type %s but item 'no' is of type %s. Please make sure they are of the same type", type2char(ta),type2char(TYPEOF(b)));
+  
+  SEXP class_type = PROTECT(getAttrib(a,R_ClassSymbol));
+  unsigned int stack_size = 1;
+  
+  // Check if same class
+  if(!R_compute_identical(class_type,getAttrib(b,R_ClassSymbol),0))
+    error("Item 'yes' has different class than item 'no'. Please make sure that they have same class type.");
+  
+  // Check if factor
+  if(isFactor(a))
+  {
+    a = PROTECT(asCharacterFactor(a));
+    b = PROTECT(asCharacterFactor(b));
```

Comment:
Afair this will make factor class always slower than character, would be nice to have it optimised where possible